### PR TITLE
TestGrid Issue-Gathering Controller: Entomologist

### PR DESCRIPTION
**Pull request by chases2**
A new controller that pins together issues and test targets
Issue #10701

### CODE REVIEW

File: testgrid/cmd/entomologist/main.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+	repository      string
+	gcsPath         string
+	gcsCredentials  string
+	oneshot         bool
+}
+
+func (o *options) parseArgs(fs *flag.FlagSet, args []string) error {
+
+	var ghCredInput string
+	fs.StringVar(&o.ghProxyEndpoint, "github-proxy", "https://api.github.com", "Address of GitHub proxy (optional)")
```
Comment (stevekuznetsov) [inline testgrid/cmd/entomologist/main.go]: Why not just GitHub endpoint? The proxy is invisible

File: testgrid/cmd/entomologist/main.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+	fs.Parse(args)
+
+	o.ghCredentials = []byte(ghCredInput)
+
+	return o.validate()
+}
+
+func (o *options) validate() error {
+	if o.organization == "" || o.repository == "" {
+		return errors.New(fmt.Sprintf("%s/%s is not a valid GitHub location", o.organization, o.repository))
```
Comment (stevekuznetsov) [inline testgrid/cmd/entomologist/main.go]: nit: use `fmt.Errorf`

Comment (stevekuznetsov) [reply testgrid/cmd/entomologist/main.go]: nit: often nicer to explicitly call out that a flag was empty -- if a user hits this they are already goofing up and `"kubernetes/ is not a valid.."` may not be clear

File: testgrid/cmd/entomologist/main.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+	fs.StringVar(&o.ghProxyEndpoint, "github-proxy", "https://api.github.com", "Address of GitHub proxy (optional)")
+	fs.StringVar(&ghCredInput, "github-token", "", "GitHub OAuth credentials (optional)")
+	fs.StringVar(&o.organization, "github-org", "", "GitHub organization")
+	fs.StringVar(&o.repository, "github-repo", "", "GitHub repository")
+	fs.StringVar(&o.gcsPath, "output", "", "GCS output (gs://bucket)")
+	fs.StringVar(&o.gcsCredentials, "gcs-cred", "", "/path/to/service/account/credentials (as .json)")
+	fs.BoolVar(&o.oneshot, "oneshot", false, "Write proto once and exit instead of monitoring GitHub for changes")
+	fs.Parse(args)
+
+	o.ghCredentials = []byte(ghCredInput)
```
Comment (stevekuznetsov) [inline testgrid/cmd/entomologist/main.go]: Why not provide the token as a file?

File: testgrid/cmd/entomologist/README.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
+# Entomologist
+
+Entomologist collects bugs from GitHub and pins them to TestGrid tests. When an issue is "pinned"
+to a TestGrid Target, a hyperlink will show up on TestGrid that leads back to the GitHub issue.
```
Comment (michelle192837) [inline testgrid/cmd/entomologist/README.md]: Can note that the hyperlink is a "(B)"

File: testgrid/cmd/entomologist/README.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+target://prow/github/report:go_default_test
+/help
+```
+
+GitHub calls Pull Requests "Issues", but Entomologist doesn't.
+Targets in Pull Requests won't be pinned.
+
+Entomologist can be configured with a caching proxy, such as [ghProxy](https://github.com/kubernetes/test-infra/tree/master/ghproxy),
+to minimize API token usage.
+Entomologist is writing to [issue_state.proto](https://github.com/kubernetes/test-infra/blob/master/testgrid/issue_state/issue_state.proto);
```
Comment (michelle192837) [inline testgrid/cmd/entomologist/README.md]: writes issue_state.proto files?

File: testgrid/cmd/entomologist/main.go (hunk 3 above)
Comment (chases2) [reply testgrid/cmd/entomologist/main.go]: GitHub's OAuth access tokens aren't very long, so it seemed better to pass them on as flags.

I'd certainly entertain a configuration file and additional options as a future enhancement. There are a few other features that I would like to include in a future draft as well.

Comment (stevekuznetsov) [reply testgrid/cmd/entomologist/main.go]: The idea behind the secret agent for long-lived processes is that it looks at a known location for secrets and re-reads the files if they change, allowing secret rotation without any intervention from the app.

File: prow/github/client.go [hunk 6]
```diff
@@ -1218,6 +1219,34 @@ func (c *client) ListIssueComments(org, repo string, number int) ([]IssueComment
...
+//
+// Each page of results consumes one API token.
+//
+// See https://developer.github.com/v3/issues/#list-issues-for-a-repository
+func (c *client) ListIssues(org, repo string) ([]Issue, error) {
+	c.log("ListIssues", org, repo)
+	if c.fake {
+		return nil, nil
+	}
+	path := fmt.Sprintf("/repos/%s/%s/issues", org, repo)
```
Comment (matthyx) [inline prow/github/client.go]: Maybe we could implement the `since` parameter to limit the number of issues to return, and save API tokens in big repos like k/k...

File: testgrid/cmd/entomologist/main.go [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+			issues = append(issues, issue)
+		}
+	}
+
+	return issues, err
+}
+
+// Algorithm for detecting and formatting test targets in issues
+func pinIssues(client githubClient, o options) (*issue_state.IssueState, error) {
+	issues, err := getIssues(client, o)
```
Comment (matthyx) [inline testgrid/cmd/entomologist/main.go]: Somehow here find a way to remember when was the last run to pass a last execution date to limit the scope of issues returned, see comment in github client.go

File: prow/github/client.go (hunk 6 above)
Comment (stevekuznetsov) [reply prow/github/client.go]: Yeah I think it's basically going to be required to run this as-is with ghproxy

Comment (chases2) [reply prow/github/client.go]: I like this idea for querying the issues, specifically.

If this sends a request, say, every hour, to k/k (~2100 issues, so ~21 tokens), then ghProxy will have to rerequest the entire 21-page resource if any issue has been added or updated in the past hour. Adding a "since" parameter might save us 20 tokens every full poll. An authorized user gets 5000 an hour, so I don't think this is necessary, but it's a nice-to-have.

We _shouldn't_ do this for the comments requests, though. Those API calls are pretty likely to be unchanged after an hour, and asking for the comments since a new timestamp is a new request, which consumes a new token. That's about 2100 tokens every hour; the real token-eater.

Comment (stevekuznetsov) [reply prow/github/client.go]: I think this hinges also on my question about why this shouldn't be a plugin. If it is, we operate incrementally and only when something changes.

File: testgrid/cmd/entomologist/main.go [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
+	logrus.Info("Entomologist is closing...")
+}
+
+var tre = regexp.MustCompile(`(?m)^target:(.+)$`)
+
+func getIssues(client githubClient, o options) ([]github.Issue, error) {
+	issuesAndPRs, err := client.ListIssues(o.organization, o.repository)
+
+	issues := issuesAndPRs[:0]
```
Comment (spiffxp) [inline testgrid/cmd/entomologist/main.go]: golang newb question: what's going on here / why is this necessary?

File: testgrid/cmd/entomologist/main.go [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+	}
+
+	return issues, err
+}
+
+// Algorithm for detecting and formatting test targets in issues
+func pinIssues(client githubClient, o options) (*issue_state.IssueState, error) {
+	issues, err := getIssues(client, o)
+
+	logrus.Infof("Found %d Issues in %s/%s", len(issues), o.organization, o.repository)
```
Comment (spiffxp) [inline testgrid/cmd/entomologist/main.go]:
```suggestion
	logrus.Infof("Found %d open issues in %s/%s", len(issues), o.organization, o.repository)
```

File: testgrid/cmd/entomologist/main.go [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+
+	if err != nil {
+		return nil, err
+	}
+
+	var results []*issue_state.IssueInfo
+
+	for _, issue := range issues {
+		newResult := issue_state.IssueInfo{
+			IssueId: strconv.FormatInt(int64(issue.Number), 10),
```
Comment (spiffxp) [inline testgrid/cmd/entomologist/main.go]: magic number: what's the `10` here?

File: testgrid/cmd/entomologist/main.go [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+
+	sigTerm := make(chan os.Signal, 1)
+	signal.Notify(sigTerm, syscall.SIGTERM)
+	signal.Notify(sigTerm, syscall.SIGINT)
+	<-sigTerm
+
+	logrus.Info("Entomologist is closing...")
+}
+
+var tre = regexp.MustCompile(`(?m)^target:(.+)$`)
```
Comment (spiffxp) [inline testgrid/cmd/entomologist/main.go]: nit: `targetRegex` or something a little more descriptive if this is going to be top-level

File: testgrid/cmd/entomologist/main.go [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+		}
+		for _, comment := range comments {
+			matches := tre.FindAllStringSubmatch(comment.Body, -1)
+			for _, match := range matches {
+				newResult.RowIds = append(newResult.RowIds, strings.TrimSpace(match[1]))
+			}
+		}
+
+		if len(newResult.RowIds) != 0 {
+			results = append(results, &newResult)
```
Comment (spiffxp) [inline testgrid/cmd/entomologist/main.go]: consider not building `newResult` until you need it down here; it looks like most of the logic above is about creating an array for `newResult.RowIds`

File: testgrid/cmd/entomologist/main_test.go [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+	}
+
+	f.IssueComments = map[int][]github.IssueComment{
+		6: {
+			{Body: "target://project:test"},
+		},
+		7: {
+			{Body: "target://project:first\r\ntarget://project:second"},
+			{Body: "target://project:third"},
+			{Body: "This is a false target: it's not at the beginning of the line"},
```
Comment (spiffxp) [inline testgrid/cmd/entomologist/main_test.go]: can you add test cases for
- `target:   leading space is trimmed`
- `target:trailing space is trimmed    ` (which apparently the markdown renderer is trimming _right now_

File: testgrid/cmd/entomologist/main.go [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+
+	client, _ := io.NewOpener(ctx, opt.gcsCredentials)
+	if client == nil {
+		logrus.Fatalf("Empty Credentials (at %s) not allowed for write operation", opt.gcsCredentials)
+	}
+
+	gcsFile := fmt.Sprintf("/bugs-%s-%s", opt.organization, opt.repository)
```
Comment (spiffxp) [inline testgrid/cmd/entomologist/main.go]: Is this filename schema arbitrary, or is this something expected by TestGrid?

**Comment by chases2** [comment]
@spiffxp Thanks for the feedback; give me some time to work through this and revise.

@stevekuznetsov The reason this isn't a prow plugin is because we're trying to separate Prow and TestGrid long-term, and having TestGrid depend on Prow for a plugin won't help. @michelle192837 can probably explain the reasons behind that effort a bit better than me.

I agree that webhooks would be a great way to handle this, and the reason I designed this initially to be [three components](https://github.com/kubernetes/test-infra/issues/10701#issuecomment-495315061) is so that we can swap out the github-interfacing part later on. It so happened that, after using the github library, all three parts were so small that they made the most sense in one go source file.

Once there's a few pieces of TestGrid interacting with GitHub, I would fully support making something `hook`-like in TestGrid to put them on, but that's a significant project.

**Comment by michelle192837** [comment]
Yeah, just to add to that; I do think having plugins in TestGrid makes some sense, but I don't think it'll work quite the same way as Prow and nothing in TestGrid is architected for that at the moment. I think eventually it will make sense, but for this particular addition I don't think we need to make it a plugin.

File: testgrid/cmd/entomologist/main.go (hunk 10 above)
Comment (chases2) [reply testgrid/cmd/entomologist/main.go]: The base of the number. If you know of an int-to-string parser that doesn't require that to be specified, let me know.

File: testgrid/cmd/entomologist/README.md [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+it's TestGrid that handles the information placed and formatted there.
+
+## Required Flags
+
+- --github-org/--github-repo: The organization/repository you're looking through for issues.\
+For example, this org/repo is `--github-org=kubernetes --github-repo=test-infra`
```
[the remark this answers is not in the document]
Comment (chases2) [reply testgrid/cmd/entomologist/README.md]: I think having a carriage return without breaking the bulleted list is important here, especially after adding backticks to the quote flags.

I'll change the indenting in the source, though

File: testgrid/cmd/entomologist/main.go [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+}
+
+func (o *options) parseArgs(fs *flag.FlagSet, args []string) error {
+
+	var ghCredInput string
+	fs.StringVar(&o.ghProxyEndpoint, "github-proxy", "https://api.github.com", "Address of GitHub proxy (optional)")
+	fs.StringVar(&ghCredInput, "github-token", "", "GitHub OAuth credentials (optional)")
+	fs.StringVar(&o.organization, "github-org", "", "GitHub organization")
+	fs.StringVar(&o.repository, "github-repo", "", "GitHub repository")
+	fs.StringVar(&o.gcsPath, "output", "", "GCS output (gs://bucket)")
```
[the remark this answers is not in the document]
Comment (chases2) [reply testgrid/cmd/entomologist/main.go]: It is, for now, required to be a bucket. A possible extension is to have it write to a local file, similar to configurator

https://github.com/kubernetes/test-infra/blob/4bd9b4592366134c54222603fe686033bd8e6a2b/testgrid/cmd/configurator/main.go#L64

I don't want to have to refactor a flag to bring TestGrid's flags more in line with each other at a future date.

File: testgrid/cmd/entomologist/main.go (hunk 8 above)
Comment (chases2) [reply testgrid/cmd/entomologist/main.go]: This initializes a new slice, "issues", with a length of 0 and the exact same data type as "issuesAndPRs"

It also creates this slice without reallocating any memory.
https://github.com/golang/go/wiki/SliceTricks#filtering-without-allocating

File: testgrid/cmd/entomologist/main.go (hunk 14 above)
Comment (chases2) [reply testgrid/cmd/entomologist/main.go]: It's expected by TestGrid, but might change in the near future

**Comment by chases2** [comment]
Regarding the extension to the github client, I had two suggestions: implement the state and since arguments, or just change the name of this function to "ListOpenIssues"

I went with the latter because it conveys an obvious function, and because we could, at a later time, write a ListIssues(org, repo, status, time string) function and have ListOpenIssues act as syntactic sugar. In general, the github client seems to only take the parameters it needs.

File: testgrid/cmd/entomologist/main_test.go (hunk 13 above)
Comment (chases2) [reply testgrid/cmd/entomologist/main_test.go]: Added in the test's comments for clarity

File: testgrid/cmd/entomologist/main.go (hunk 12 above)
Comment (chases2) [reply testgrid/cmd/entomologist/main.go]: I have done so, and renamed the slice of targets this routine is building to "targets"

File: testgrid/cmd/entomologist/README.md [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+
+## Pinning Behavior
+
+Entomologist is expecting targets to be explicitly called out in issues. You can do this by writing
+"target:", and then the target you want to pin to.
+
+```
+Some update caused these tests to start failing!
+
+target://prow/git:go_default_test
```
[the remark this answers is not in the document]
Comment (chases2) [reply testgrid/cmd/entomologist/README.md]: I've used a quotation in this revision. I want to convey clearly that this is an example you can type to make this feature work, and I think both backtick and quotation blocks do that.

I'd like feedback on which you prefer

File: prow/github/client.go [hunk 18]
```diff
@@ -1218,6 +1219,34 @@ func (c *client) ListIssueComments(org, repo string, number int) ([]IssueComment
 	return comments, nil
 }
 
+// ListIssues returns all open issues, including pull requests
+//
+// Each page of results consumes one API token.
+//
+// See https://developer.github.com/v3/issues/#list-issues-for-a-repository
+func (c *client) ListIssues(org, repo string) ([]Issue, error) {
```
[the remark this answers is not in the document]
Comment (chases2) [reply prow/github/client.go]: I've renamed this, and will consider "state" for a future "ListIssues" function, if needed.

I talk a bit more about this decision with [this comment](https://github.com/kubernetes/test-infra/pull/12885#issuecomment-499703622)

File: testgrid/cmd/entomologist/main.go (hunk 10 above)
Comment (stevekuznetsov) [reply testgrid/cmd/entomologist/main.go]: `strconv.Itoa`?

File: testgrid/cmd/entomologist/README.md (hunk 4 above)
Comment (chases2) [reply testgrid/cmd/entomologist/README.md]: A target is a particular test that TestGrid is displaying. This is usually a row.

I've modified the readme to try to explain this implicitly.

Comment (chases2) [reply testgrid/cmd/entomologist/README.md]: A target is a particular test that TestGrid is reporting. This is usually a row.

I've changed the readme to try to explain this implicitly

File: testgrid/cmd/entomologist/main.go [hunk 19]
```diff
@@ -0,0 +1,224 @@
...
+type githubClient interface {
+	ListOpenIssues(org, repo string) ([]github.Issue, error)
+	ListIssueComments(org, repo string, number int) ([]github.IssueComment, error)
+}
+
+const defaultPollTime = 1 * time.Hour
+
+type options struct {
+	//ghProxyEndpoint string
+	//ghCredentials   []byte
```
Comment (spiffxp) [inline testgrid/cmd/entomologist/main.go]: I would drop the commented out code

File: testgrid/cmd/entomologist/main.go [hunk 20]
```diff
@@ -0,0 +1,224 @@
...
+	repository     string
+	gcsPath        string
+	gcsCredentials string
+	oneshot        bool
+}
+
+func (o *options) parseArgs(fs *flag.FlagSet, args []string) error {
+
+	//fs.StringVar(&o.ghProxyEndpoint, "github-proxy", "https://api.github.com", "Address of GitHub proxy (optional)")
+	//fs.StringVar(&ghCredInput, "github-token", "", "GitHub OAuth credentials (optional)")
```
Comment (spiffxp) [inline testgrid/cmd/entomologist/main.go]: same

File: testgrid/cmd/entomologist/main.go (hunk 14 above)
Comment (spiffxp) [reply testgrid/cmd/entomologist/main.go]: OK, in that case can we leave a comment explaining as much? eg: 
```
// testgrid expects issue_state.proto files for org/repo to be be named bugs-org-repo
```

File: testgrid/cmd/entomologist/main.go [hunk 21]
```diff
@@ -0,0 +1,224 @@
...
+	for _, issue := range issues {
+
+		var targets []string
+
+		matches := targetRegExp.FindAllStringSubmatch(issue.Body, -1)
+		for _, match := range matches {
+			targets = append(targets, strings.TrimSpace(match[1]))
+		}
+
+		// todo(chases2): separate the comments API calls into their own goroutines
```
Comment (spiffxp) [inline testgrid/cmd/entomologist/main.go]: nit
```suggestion
		// TODO(chases2): separate the comments API calls into their own goroutines
```

File: testgrid/cmd/entomologist/main.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (spiffxp) [reply testgrid/cmd/entomologist/main.go]: your validation is only enforcing that this start with `gs://`, which to me implies it could accept a full path such as `gs://bucket/some/path/to/a/dir`

if this is required to be a bucket, you'd want to validate that it's exactly `gs://bucketname`

**Comment by chases2** [comment]
Note: If this merges after or during #12945 merges, I'll have to make a change to how the github fake works.

We're both dealing with issues, and that change turns the fake from a slice to a map

File: testgrid/cmd/entomologist/main.go (hunk 14 above)
Comment (chases2) [reply testgrid/cmd/entomologist/main.go]: Done

**Comment by chases2** [comment]
@spiffxp I'm not forcing that validation because, while it's not usually what you would do, there's no reason TestGrid can't be configured to pull from a subdirectory instead of a bucket. Both my comments and the little help blurb implied that it _must_ be a bucket, and that's not quite right; it only probably is.

I've also refactored the testing to make the test cases run in parallel and independently of each other, instead of as one "test-repository". The exact same cases are being tested.

**Comment by chases2** [comment]
The merge conflict is resolved with this version
